documentation/specs/task-isolation-and-dependencies.md

```diff
@@ -0,0 +1,38 @@
+# Task isolation
+## Problem definition
+Tasks in MSBuild are dynamically loaded assemblies with potentially separate and colliding dependency trees. Currently MSBuild on .NET Core has no isolation between tasks and as such only one version of any given assembly can be loaded. Prime example of this is Newtonsoft.Json which has multiple versions, but all the tasks must agree on it to work.
+This problem is also described in microsoft/msbuild#1754.
+
+## Possible solution
+Use [`AssemblyLoadContext`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.loader.assemblyloadcontext?view=netcore-2.2) (ALC) to provide binding isolation for tasks. Each task would be loaded into its own ALC instance.
```

Comment:
Should this be at task (class) or assembly granularity? The whole assembly would have to have the same dependency closure, right?

Reply:
Definitely assembly level - I'll change the wording..